Improved notebook; Spot Instance Support; Code for single/multi-CPU/GPU

miroenev opened the pull request:
- Improved notebook text and figures
- Combined all key choices in preamble 
[ dataset size, code type, algorithm choice, cv folds, max_jobs, max_parallel_jobs, search strategy, HPO ranges ]
- Added spot instance support
- Added instance recommendation function based on configuration choices
- Added code branches for 
   * `single-CPU` ( pandas + sklearn )
   * `multi-CPU` ( dask + pandas + sklearn )
   * `single-GPU` ( cudf + cuml )
   * `multi-GPU` ( dask + cudf + cuml )

JohnZed commented on aws/README.md:
```diff
@@ -27,4 +27,4 @@
    > Note: If you see Pending to the right of the notebook instance in the Status column, your notebook is still being created. The status will change to InService when the notebook is ready for use.
 
 3. Run Notebook
```
Maybe note somewhere that it won't work in Studio currently?

miroenev replied:
Done

JohnZed commented on aws/code/rapids_cloud_ml.py:
```diff
@@ -15,289 +15,450 @@
+                'max_depth' : args.max_depth,
+                'n_estimators' : args.n_estimators,        
+                'max_features': args.max_features,
+                'seed' : 0,
```
Would be nice to allow users to pass in a fixed seed.

miroenev replied:
Done

JohnZed commented on aws/code/rapids_cloud_ml.py:
```diff
@@ -15,289 +15,450 @@
+    # -------------------------------------------------------------------------------------------------------------
+    # ETL
+    # -------------------------------------------------------------------------------------------------------------
+    airline_label_column = 'ArrDel15'
```
Should we separate out the airline-specific code a bit more from the independent code?

miroenev replied:
Good idea, moved dataset feature and label column definitions to train.py. Hopefully this encourages more BYOData uses.

JohnZed commented on aws/notebooks/helper_functions.py:
```diff
@@ -0,0 +1,77 @@
+def new_job_name_from_config ( dataset_directory, code_choice, 
+                               algorithm_choice, cv_folds,
+                               instance_type, trim_limit = 32 ):
+    job_name = None    
```
It would be nice to have docstrings since these are user accessible

miroenev replied:
Done

zronaghi commented on aws/notebooks/hpo.ipynb:
```diff
@@ -0,0 +1,1047 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "max_jobs = 100"
```
Probably better if default is less than 100, can add a comment for larger jobs

miroenev replied:
Done

miroenev commented on aws/README.md:
```diff
@@ -27,4 +27,4 @@
 
 3. Run Notebook
-   - Once inside JupyterLab you should be able to navigate to the notebook in the **aws/notebooks** directory named **sagemaker_rapids_hpo.ipynb**
+   - Once inside JupyterLab you should be able to navigate to the notebook in the **aws/notebooks** directory named **hpo.ipynb**
```
I'll go back to the old name, I was thinking it may be overly verbose but it does add redundancy/clarity.

miroenev commented on aws/code/rapids_cloud_ml.py:
```diff
@@ -15,289 +15,450 @@
+        else:
+            target_files = dataset_structure['train_data'] + str(dataset_path)
+                
+        n_datafiles = len( glob.glob( target_files ) )
```
This was more of a sanity check, to make sure that the bucket you've pointed to is holding the data you are expecting. I can add this additional context as a comment.

Are you OK with keeping this check in the code?

miroenev commented:
- Finished large XGBoost run and added acceleration + cost savings factors
- Added cluster teardown between folds [ limit memory creep for high CV counts]
- Removed dependence on airline dataset from ETL code [ moved to train.py ]
- Changed RF model save/persistence to use joblib.dump
- Added docstrings
- Added comment in README.md regarding lack of a SageMaker Studio version
- Added seed for XGBoost and RandomForest parsable parameter

zronaghi commented:
Thanks, looks good to me!
